### PR TITLE
Issue120 121

### PR DESCRIPTION
**Pull request by meta00**
This pull request addresses the issue:

#120: to allow constructing an experiment object with no information, except link to a gaml file.
#121 : to check tmax (either NULL or positive integer), seed (either NULL or integer, gama doesn't run with a double seed?) @gnoubi 
#118: to force type of seed, tmax and obsrates from numeric to integer.

### CODE REVIEW

File: R/experiment.R [hunk 1]
```diff
@@ -124,23 +150,35 @@ validate_experiment <- function(x) {
   colnames <- lapply(c(parameters, obs_rates), function(f) names(f(x)))
 
   check_experiment(name(x), model)
-  test_schar(names(dic_g2r))
+  # test_schar(names(dic_g2r)) check if it works in different gama versions
```
Comment (LucieContamin) [inline R/experiment.R]: We can remove it as for the display we need the last version of GAMA and in the last version, we don't have this problem of special character in the name of parameters.

**Comment by LucieContamin** [comment]
I just want to add here, to be clear, if we remove the casting `as.numeric` in the functions `get_parameters` and `get_variables` of the file `load_experiment`, it's because we are doing it in the function `new_experiment` called by the function `experiment` used in `load_experiment`.

File: R/experiment.R [hunk 2]
```diff
@@ -18,67 +18,93 @@ new_experiment <- function(parameters, obsrates, tmax, seed,
...
+    dic_r2g <- setNames(names(dic_g2r), dic_g2r)
   }
 
-  # Dealing with obsrates and tmax, converting them into integers if needed:
-  if (any(!sapply(obsrates, is.integer))) {
+  # Dealing with obsrates, seed and tmax, converting them into integers if needed:
+  if (!all(sapply(obsrates, is.integer))) {
     message(cat(
       "Periods of observation (\"obsrates\") are rounded and converted into",
       " integers."))
```
[the remark this answers is not in the document]
Comment (meta00) [reply R/experiment.R]: Yes, agree

File: R/experiment.R [hunk 3]
```diff
@@ -18,67 +18,93 @@ new_experiment <- function(parameters, obsrates, tmax, seed,
...
     message(cat(
       "Final time step (\"tmax\") is rounded and converted into integer."))
     tmax <- as.integer(tmax)
   }
 
-  model_info <- list("path" = model,
+  # seed can't be double in gama?, to be confirmed
+  if (!all(is.integer(seed))) {
+    message(cat(
+      "Seed is rounded and converted into integer."))
```
[the remark this answers is not in the document]
Comment (meta00) [reply R/experiment.R]: Yes, agree

File: R/map_experiment.R [hunk 4]
```diff
@@ -24,21 +24,44 @@
 map_experiment <- function(df, exp) {
   stopifnot(is.data.frame(df))
 
+  # Type casting for numeric values, character not passed
+  obsrates <- names(df)[grepl("r_", names(df))]
+  if (!all(sapply(df[, obsrates], is.numeric))) {
+    message(cat(
+      "Periods of observation (\"obsrates\") are rounded and converted into",
+      " integers."))
```
[the remark this answers is not in the document]
Comment (meta00) [reply R/map_experiment.R]: Yes, agree

File: R/experiment.R [hunk 5]
```diff
@@ -18,67 +18,93 @@ new_experiment <- function(parameters, obsrates, tmax, seed,
   #   data.frame(S = 1, I = 1, R = 1),
   #   1000, 1, "sir", system.file("models", "sir.gaml", package = "rama"))
 
-  stopifnot(is.data.frame(parameters))
-  stopifnot(is.data.frame(obsrates))
-  stopifnot(is.numeric(tmax))
-  stopifnot(is.null(dim(seed)))
```
[the remark this answers is not in the document]
Comment (meta00) [reply R/experiment.R]: True

[the remark this answers is not in the document]
Comment (LucieContamin) [reply R/experiment.R]: Seed can be NULL so the test is not necessary anymore

File: R/experiment.R [hunk 6]
```diff
@@ -18,67 +18,93 @@ new_experiment <- function(parameters, obsrates, tmax, seed,
...
-  stopifnot(is.character(experiment))
-  stopifnot(length(experiment) == 1)
```
[the remark this answers is not in the document]
Comment (LucieContamin) [reply R/experiment.R]: Done later in the code line 31

File: R/experiment.R [hunk 7]
```diff
@@ -18,67 +18,93 @@ new_experiment <- function(parameters, obsrates, tmax, seed,
...
-  stopifnot(is.character(model))
-  stopifnot(length(model) == 1)
```
[the remark this answers is not in the document]
Comment (LucieContamin) [reply R/experiment.R]: Done later in the code, line 49

File: R/experiment.R (hunk 2 above)
[the remark this answers is not in the document]
Comment (LucieContamin) [reply R/experiment.R]: Done in commit `9603784`

File: R/experiment.R [hunk 8]
```diff
@@ -18,67 +18,93 @@ new_experiment <- function(parameters, obsrates, tmax, seed,
...
-    obsrates[] <- lapply(obsrates, function(x) as.integer(round(x)))
+    obsrates[] <- lapply(obsrates, function(x) as.integer(x))
   }
-  if (!is.numeric(tmax))
-    stop(cat("\"tmax\" must have a numeric value"))
-  if (!is.integer(tmax)) {
+
+  if (!all(is.integer(tmax))) {
     message(cat(
       "Final time step (\"tmax\") is rounded and converted into integer."))
```
[the remark this answers is not in the document]
Comment (LucieContamin) [reply R/experiment.R]: Done in commit `9603784`

File: R/experiment.R (hunk 3 above)
[the remark this answers is not in the document]
Comment (LucieContamin) [reply R/experiment.R]: Done in commit `9603784`

[the remark this answers is not in the document]
Comment (LucieContamin) [reply R/experiment.R]: Done in commit `9603784`

File: R/map_experiment.R (hunk 4 above)
[the remark this answers is not in the document]
Comment (LucieContamin) [reply R/map_experiment.R]: Done in commit `9603784`

File: R/map_experiment.R [hunk 9]
```diff
@@ -24,21 +24,44 @@
...
+
   if (setequal(names(df), names(exp))) {
     the_attributes <- attributes(exp)
     the_attributes$row.names <- row.names(df)
     attributes(df) <- the_attributes
     return(validate_experiment(df))
   }
 
-  if (all(names(df) %in% names(exp)) & nrow(df) == nrow(exp)) {
+  if (any(names(df) %in% names(exp), nrow(df) == nrow(exp))) {
```
[the remark this answers is not in the document]
Comment (LucieContamin) [reply R/map_experiment.R]: Done in commit `ee6372b`

File: R/experiment.R (hunk 1 above)
Comment (LucieContamin) [reply R/experiment.R]: Done in commit `2f48cdf`

**Comment by LucieContamin** [comment]
This pull-request resolve also the issue #125

**Pull request merged**